[UNIT] Final review of Unit 2.1 on smolagents - release 25/2/25 (#195)
huggingface/agents-course: units/en/unit2/smolagents/multi_agent_systems.md

# Multi-Agent Systems

Multi-agent systems enable specialized agents to collaborate on complex tasks, improving modularity, scalability, and robustness. Instead of relying on a single agent, tasks are distributed among agents with distinct capabilities.

> In **smolagents**, different agents can be combined to generate Python code, call external tools, perform web searches, and more. By orchestrating these agents, we can create powerful workflows.

Reviewer: ```suggestion
In `smolagents`, different agents can be combined to generate Python code, call external tools, perform web searches, and more. By orchestrating these agents, we can create powerful workflows.
```

A typical setup might include:
- A **Manager Agent** for task delegation  
- A **Code Interpreter Agent** for code execution  
- A **Web Search Agent** for information retrieval  

> The diagram below illustrates a simple multi-agent architecture where a **Manager Agent** coordinates a **Code Interpreter Tool** and a **Web Search Agent**, which in turn utilizes tools like `Web Search` and `Visit Webpage` to gather relevant information.

Reviewer: ```suggestion
The diagram below illustrates a simple multi-agent architecture where a **Manager Agent** coordinates a **Code Interpreter Tool** and a **Web Search Agent**, which in turn utilizes tools like the `DuckDuckGoSearchTool` and `VisitWebpageTool` to gather relevant information.
```

<img src="https://mermaid.ink/img/pako:eNp1kc1qhTAQRl9FUiQb8wIpdNO76eKubrmFks1oRg3VSYgjpYjv3lFL_2hnMWQOJwn5sqgmelRWleUSKLAtFs09jqhtoWuYUFfFAa6QA9QDTnpzamheuhxn8pt40-6l13UtS0ddhtQXj6dbR4XUGQg6zEYasTF393KjeSDGnDJKNxzj8I_7hLW5IOSmP9CH9hv_NL-d94d4DVNg84p1EnK4qlIj5hGClySWbadT-6OdsrL02MI8sFOOVkciw8zx8kaNspxnrJQE0fXKtjBMMs3JA-MpgOQwftIE9Bzj14w-cMznI_39E9Z3p0uFoA?type=png" style='background: white;'>

## Multi-Agent Systems in Action  

A multi-agent system consists of multiple specialized agents working together under the coordination of an **Orchestrator Agent**. This approach enables complex workflows by distributing tasks among agents with distinct roles.  

For example, a **Multi-Agent RAG system** can integrate:  
- A **Web Agent** for browsing the internet.  
- A **Retriever Agent** for fetching information from knowledge bases.  
- An **Image Generation Agent** for producing visuals.  

All of these agents operate under an orchestrator that manages task delegation and interaction.  

> # Solving a complex task with a multi-agent hierarchy

Reviewer: ```suggestion
## Solving a complex task with a multi-agent hierarchy
```

model = OpenAIServerModel(model_id="gpt-4o")
```

We can start with creating a baseline, simple agent to give us a simple report.

```python
task = """Find all Batman filming locations in the world, calculate the time to transfer via cargo plane to here (we're in Gotham, 40.7128° N, 74.0060° W), and return them to me as a pandas dataframe.
Also give me some supercar factories with the same cargo plane transfer time."""
```

```python
import pandas as pd

agent = CodeAgent(
    model=model,
    tools=[DuckDuckGoSearchTool(), VisitWebpageTool(), calculate_cargo_travel_time],
    additional_authorized_imports=["pandas"],
    max_steps=20,
)

result = agent.run(task)result
```

We could already improve this a bit by throwing in some dedicated planning steps, and adding more prompting.

```python
agent.planning_interval = 4

detailed_report = agent.run(f"""
You're an expert analyst. You make comprehensive reports after visiting many websites.
Don't hesitate to search for many queries at once in a for loop.
For each data point that you find, visit the source url to confirm numbers.

{task}
""")

print(detailed_report)
```

> ```python
> detailed_report
> ```
> 

Reviewer: ```suggestion
```

Reviewer: Or does this show some kind of information? Perhaps we should then mention this above.

Thanks to these quick changes, we obtained a much more concise report by simply providing our agent a detailed prompt, and giving it planning capabilities!

💸 But as you can see, the context window is quickly filling up. So **if we ask our agent to combine the results of detailed search with another, it will be slower and quickly ramp up tokens and costs**.

➡️ We need to improve the structure of our system.

> ## ✌️ Splitting the task between two agents

Reviewer: ```suggestion

### ✌️ Splitting the task between two agents
```

Multi-agent structures allow to separate memories between different sub-tasks, with two great benefits:
- Each agent is more focused on its core task, thus more performant
- Separating memories reduces the count of input tokens at each step, thus reducing latency and cost.

Let's create a team with a dedicated web search agent, managed by another agent.

The manager agent should have plotting capabilities to redact its final report: so let us give it access to additional imports, including `matplotlib`, and `geopandas` + `shapely` for spatial plotting.

> ```python
> !pip install matplotlib geopandas shapely -q
> 
> model = OpenAIServerModel("gpt-4o")
> 
> web_agent = CodeAgent(
>     model=model,
>     tools=[DuckDuckGoSearchTool(), VisitWebpageTool(), calculate_cargo_travel_time],
>     name="web_agent",
>     description="Browses the web to find information",
>     verbosity_level=1,
>     max_steps=10,
> )
> ```

Reviewer: ```suggestion
Let's first install the required packages.

```python
!pip install matplotlib geopandas shapely -q
```

Now, we can start initialise our agent.

```python
model = OpenAIServerModel("gpt-4o")

web_agent = CodeAgent(
    model=model,
    tools=[DuckDuckGoSearchTool(), VisitWebpageTool(), calculate_cargo_travel_time],
    name="web_agent",
    description="Browses the web to find information",
    verbosity_level=1,
    max_steps=10,
)
```
```

Reviewer: Sshould we add more import here too?

            └─────────────────────────────┴───────────────────────────────────┴───────────────────────────────────┘
```

```python
manager_agent.run(f"""
{task}
Represent this as spatial map of the world, with the locations represented as scatter points proportional in size to the travel time, and save it to saved_map.png!

Here's an example of how to plot and return a map:
import plotly.express as px
df = px.data.carshare()
fig = px.scatter_map(df, lat="centroid_lat", lon="centroid_lon", text="name", color="peak_hour", size="car_hours",
     color_continuous_scale=px.colors.sequential.Magma, size_max=15, zoom=10)
fig.show()
fig.write_image("saved_image.png")
final_answer(fig)
""")
```

I don't know how that went in your run, but in mine, the manager agent masterfully divided tasks given to the web agent in `1. Search for Batman filming locations`, then `2. Find supercar factories`, before aggregating the lists and plotting the map.

Let's see what the map looks like by inspecting it directly from the agent state:

```python
manager_agent.python_executor.state["fig"]
```

> ## Adding telemetry

Reviewer: Some parts are missing/unfinished here?

Author: That's correct. We have this PR https://github.com/huggingface/agents-course/pull/199#issue-2863732774
